internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher_test.go

```diff
@@ -6,74 +6,1205 @@ package githubupstreamwatcher
+				{
+					Name:               "some-idp-name",
+					ResourceUID:        "some-resource-uid",
+					Host:               fmt.Sprintf("https://%s", *validFilledOutIDP.Spec.GitHubAPI.Host),
+					UsernameAttribute:  "id",
+					GroupNameAttribute: "name",
+					OAuth2Config: &oauth2.Config{
+						ClientID:     "some-client-id",
+						ClientSecret: "some-client-secret",
+					},
+					AllowedOrganizations:    []string{"organization1", "org2"},
+					OrganizationLoginPolicy: "OnlyUsersFromAllowedOrganizations",
+					AuthorizationURL:        fmt.Sprintf("https://%s/login/oauth/authorize", *validFilledOutIDP.Spec.GitHubAPI.Host),
+					HttpClient:              buildPretendHttpClient(t, goodServer),
+				},
+			},
+			wantResultingUpstreams: []v1alpha1.GitHubIdentityProvider{
+				{
+					ObjectMeta: validFilledOutIDP.ObjectMeta,
+					Spec:       validFilledOutIDP.Spec,
+					Status: v1alpha1.GitHubIdentityProviderStatus{
+						Phase: v1alpha1.GitHubPhaseReady,
+						Conditions: []metav1.Condition{
+							buildClientCredentialsObtainedTrue(t),
+							buildGitHubConnectionValidTrue(t, *validFilledOutIDP.Spec.GitHubAPI.Host),
+							buildHostValid(t, metav1.ConditionTrue, *validFilledOutIDP.Spec.GitHubAPI.Host),
+							buildOrganizationsPolicyValidTrue(t),
+							buildTLSConfigurationValid(t, metav1.ConditionTrue, *validFilledOutIDP.Spec.GitHubAPI.Host),
+						},
+					},
+				},
+			},
+		},
+		{
+			name:    "happy path with minimal fields",
+			secrets: []runtime.Object{goodSecret},
+			githubIdentityProviders: []runtime.Object{
+				validMinimalIDP,
+			},
+			wantResultingCache: []*upstreamgithub.ProviderConfig{
+				{
+					Name:               "minimal-idp-name",
+					ResourceUID:        "minimal-uid",
+					Host:               fmt.Sprintf("https://%s", goodServerDomain),
+					UsernameAttribute:  "login",
+					GroupNameAttribute: "slug",
+					OAuth2Config: &oauth2.Config{
+						ClientID:     "some-client-id",
+						ClientSecret: "some-client-secret",
+					},
+					OrganizationLoginPolicy: "AllGitHubUsers",
+					AuthorizationURL:        fmt.Sprintf("https://%s/login/oauth/authorize", goodServerDomain),
+					HttpClient:              buildPretendHttpClient(t, goodServer),
+				},
+			},
+			wantResultingUpstreams: []v1alpha1.GitHubIdentityProvider{
+				{
+					ObjectMeta: validMinimalIDP.ObjectMeta,
+					Spec:       validMinimalIDP.Spec,
+					Status: v1alpha1.GitHubIdentityProviderStatus{
+						Phase: v1alpha1.GitHubPhaseReady,
+						Conditions: []metav1.Condition{
+							buildClientCredentialsObtainedTrue(t),
+							buildGitHubConnectionValidTrue(t, *validMinimalIDP.Spec.GitHubAPI.Host),
+							buildHostValid(t, metav1.ConditionTrue, *validMinimalIDP.Spec.GitHubAPI.Host),
+							buildOrganizationsPolicyValidTrue(t),
+							buildTLSConfigurationValid(t, metav1.ConditionTrue, *validMinimalIDP.Spec.GitHubAPI.Host),
+						},
+					},
+				},
+			},
+		},
+		{
+			name: "multiple idps - two good, one invalid",
```

Comment:
I tend to prefer the `name` to contain `will do X,Y,Z`, so that if it fails later, its easy to understand what the behavior was supposed to be (and decide if we made a mistake and changed behavior, or if we intentionally changed behavior).

Reply:
That's fair, in this case that's really hard because I wanted to make sure the three objects were each treated independently by the controller

Reply:
I think the only thing that really concerns me is if it "should enueue a resync" or not.  Agree, we know the conditions should always update.